src/metadata.rs

```diff
@@ -25,30 +25,165 @@ pub struct Metadata {
     imdb: Option<String>,
 }
 
+macro_rules! check_pattern {
+    ($pattern:ident, $name:ident, $title_start:ident, $title_end:ident, $get_capture:expr) => {
+        $pattern
+            .captures($name)
+            .map(|caps| {
+                if let Some(cap) = caps.get(0) {
+                    if $pattern.before_title() {
+                        $title_start = max($title_start, cap.end());
+                    } else {
+                        $title_end = min($title_end, cap.start());
+                    }
+                };
+                $get_capture(caps)
+            })
+            .flatten()
+    };
+    ($pattern:ident, $name:ident, $title_start:ident, $title_end:ident) => {
+        $pattern.captures($name).map(|caps| {
+            if let Some(cap) = caps.get(0) {
+                if $pattern.before_title() {
+                    $title_start = max($title_start, cap.end());
+                } else {
+                    $title_end = min($title_end, cap.start());
+                }
+            };
+            caps
+        })
+    };
+}
+
+macro_rules! match_to_string {
+    ($name:ident) => {
+        $name.and_then(|c| c.get(0)).map(|m| m.as_str().to_string())
+    };
+}
+
 impl Metadata {
     pub fn from(name: &str) -> Result<Self, ErrorMatch> {
         let mut title_start = 0;
         let mut title_end = name.len();
 
-        let patterns = all_patterns().collect::<Vec<_>>();
-        let mut captures = HashMap::with_capacity(patterns.len());
-        for (pname, p) in patterns {
-            if let Some(m) = p.captures(name) {
-                if let Some(cap) = m.get(0) {
-                    if p.before_title() {
-                        title_start = max(title_start, cap.end());
-                    } else {
-                        title_end = min(title_end, cap.start());
-                    }
-                }
-                captures.insert(*pname, m);
-            }
-        }
+        let season: Option<i32> = check_pattern!(
+            PATTERN_SEASON,
+            name,
+            title_start,
+            title_end,
+            |caps: Captures<'_>| caps
+                .name("short")
+                .or_else(|| caps.name("long"))
+                .or_else(|| caps.name("dash"))
+                .map(|m| m.as_str())
+                .map(|s| s.parse().unwrap())
```

Comment:
This and the later ones can probably be done in a single map

Reply:
This doesn't work with integer primitives (they `impl ToString` but there's no `From` impl on `String` for them), but I changed it to defer parsing until after the error check anyway (bonus: only allocate if there is an error), so this works